cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,348 @@ struct whitespace_rsplit_tokenizer_fn
+      }
+      offset_buf_ptr[0] = 0;
+    }
+    offset_buf_ptr[token_count] = token_size_sum;
+  }
+};
+
+/**
+ * @brief Compute the number of tokens, the total byte sizes of the tokens, and
+ * required memory size for the `idx'th` string element of `d_strings`.
+ */
+template <bool forward>
+struct whitespace_token_reader_fn {
+  column_device_view const d_strings;  // strings to split
+  size_type const max_tokens = std::numeric_limits<size_type>::max();
+  bool const has_validity = false;
+
+  __device__ thrust::tuple<size_type, size_type, size_type>
+  operator()(size_type idx) const {
+    if (has_validity && d_strings.is_null(idx)) {
+      return thrust::make_tuple<size_type, size_type, size_type>(0, 0, 0);
+    }
+
+    auto d_str = d_strings.element<string_view>(idx);
+    size_type token_count = 0;
+    size_type token_size_sum = 0;
+    auto spaces = true;
+    auto reached_max_tokens = false;
+    size_type to_token_pos = 0;
+    for (size_type i = 0; i < d_str.length(); ++i) {
+      auto cur_pos = forward ? i : d_str.length() - 1 - i;
+      auto ch = d_str[cur_pos];
+      if (spaces != (ch <= ' ')) {
+        if (spaces) {  // from whitespace(s) to a new token
+          to_token_pos = cur_pos;
+        }
+        else {  // from a token to whiltespace(s)
+          if (token_count < max_tokens - 1) {
+            token_count++;
+            token_size_sum +=
+              forward ?
+                d_str.byte_offset(cur_pos) - d_str.byte_offset(to_token_pos) :
+                d_str.byte_offset(to_token_pos + 1) -
+                  d_str.byte_offset(cur_pos + 1);
+          }
+          else {
+            reached_max_tokens = true;
+            break;
+          }
+        }
+        spaces = !spaces;
+      }
+    }
+    if (reached_max_tokens || !spaces) {
+      token_count++;
+      token_size_sum +=
+        forward ?
+          d_str.byte_offset(d_str.length()) - d_str.byte_offset(to_token_pos) :
+          d_str.byte_offset(to_token_pos + 1) - d_str.byte_offset(0);
+    }
+
+    if (token_count == 0) {  // note that pandas.Series.str.split("", pat=" ")
+                             // returns one token (i.e. "") while
+                             // pandas.Series.str.split("") returns 0 token.
+      return thrust::make_tuple<size_type, size_type, size_type>(0, 0, 0);
+    }
+
+    auto memory_size =
+      round_up_pow2(token_size_sum, split_align) +
+      round_up_pow2(
+        (token_count + 1) * static_cast<size_type>(sizeof(size_type)),
+        split_align);
+
+    return thrust::make_tuple<size_type, size_type, size_type>(
+      token_count, token_size_sum, memory_size);
+  }
+};
+
+/**
+ * @brief Copy the tokens from the `idx'th` string element of `d_strings` to
+ * the contiguous memory buffer.
+ */
+template <bool forward>
+struct whitespace_token_copier_fn {
+  column_device_view const d_strings;  // strings to split
+  bool const has_validity = false;
+
+  __device__ void
+  operator()(
```

Comment:
Consider refactoring operator() into separate calls to several __device__ member functions for readability. Perhaps the loops might be good candidates to factor out.